### DISCUSSION TITLE
Interpolated Color Gradient across Drawable.Line

### DISCUSSION THREAD

**Discussion by Shook-Software** [dlemstra/Magick.NET#1298] [Help]
Hey, I'm trying to interpolate colors across each line I draw using Drawable.Line() but I can't seem to find any way to apply such an operation on a single line, only upon the full image. Each line has a given color, and I want to interpolate from a darkened color at each end of the line to the given color in the middle.

Am I mistaken in my approach, or is there an operation to do just this that I'm not seeing?

Thanks!

**Comment by dlemstra** [comment]
Can you explain visually what you are trying to do? Not sure if I understand what you want.

**Reply by Shook-Software** [reply]
![NewTest](https://user-images.githubusercontent.com/113054672/201986880-633a9981-13a9-4192-95d0-41e2b8b75d2a.png)
![OriginalTest](https://user-images.githubusercontent.com/113054672/201986893-2dfa8245-14c1-49ff-8b8e-b4c5edec3e0b.png)

Thanks for the quick reply.
The first image is what I have so far using ImageMagick.
The second image has the edges of each line darkened, and was made using System.Drawing.
I'm using ImageMagick to create logos from an embroidery file without the reliance on a windows operating system, where System.Drawing is not portable, so it is being replaced. 
So, for each line, it should have a gradient of dark to light to dark again for whichever color is being drawn.

**Reply by Shook-Software** [reply]
Is more information needed?

**Reply by dlemstra** [reply]
I haven't had time to take a look at this. I have a lot of things that need my attention so I don't know when I will be able to help you with this. You could ask your question on the ImageMagick discussions and if you have a command line example you will then need to translate it to C#.  And it would also be a good idea to share your input image?

**Reply by Shook-Software** [reply]
Coming back to this, I am still struggling to create lines with a gradient across them. I will try to explain the overarching problem, and it should be noted my experience with image processing is minimal.

There is no input image, as each line is drawn onto a transparent image produced with new MagickImage("xc:transparent", width, height).

I am reading information from an embroidery file (.DST extension). The embroidery file allows me to determine stitches (straight lines) that have a color stretching from one coordinate (x1, y1) to another (x2, y2).

Two resultant images are shown above. 
The first is using the  ImageMagick Line drawable. 
The second is using a Windows specific library, however we want to be able to port this code to any system and put it in containers. Hence the change to ImageMagick. This second image has gradients on each line that comprises the overall image of an embroidered logo.

There are roughly 3700 stitches, or 3700 lines in both images.

I have managed to create a singular line with the desired gradient effect by composing two separate images colored using image.SparseColor.

```
int startX = 50;
int startY = 50;

int endX = 200;
int endY = 200;

double angle = Math.Atan2(endY - startY, endX - startX) * 180 / Math.PI;
double absoluteLength = Math.Sqrt(Math.Pow((endX - startX), 2) + Math.Pow(endY - startY, 2));

MagickColor darkerRed = new((byte)(MagickColors.Red.R * 0.2), (byte)(MagickColors.Red.G * 0.2), (byte)(MagickColors.Red.B * 0.2));

var sparseColorsStart = new SparseColorArg[]
{
    new(0, 0, darkerRed),
    new(absoluteLength / 2, 0, MagickColors.Red)
};

var sparseColorsEnd = new SparseColorArg[]
{
    new(0, 0, MagickColors.Red),
    new(absoluteLength / 2, 0, darkerRed)
};

MagickImage line1 = new("xc:transparent", (int)absoluteLength, 3);
line1.SparseColor(SparseColorMethod.Bilinear, sparseColorsStart);

MagickImage line2 = new("xc:transparent", (int)absoluteLength / 2, 3);
line2.SparseColor(SparseColorMethod.Bilinear, sparseColorsEnd);

line1.Composite(line2, (int)absoluteLength / 2, 0);
//line1.Rotate(angle); //not needed for base example

line1.Write("filePathOmitted");
```

The produced image is:
![TestLine](https://user-images.githubusercontent.com/113054672/207154335-7c6e4e2f-ada7-4625-9839-b5f08e95881c.png)

However, this is slow and trickier when applied across the sheer number of required compositions to produce the final image.
Therefore, using the Drawables is ideal, but I cannot figure out how to achieve any sort of gradient on a Drawable. Is this something the library or underlying command line functions support? Am I totally missing the mark here?
I'd appreciate any thoughts on the matter.
